### PR TITLE
Implement the nullability axis for jkinds

### PR DESCRIPTION
**Pull request by dkalinichenko-js**
Implement the `Nullability` axis for jkinds, with `Non_null < Or_null`.

Add legacy layouts `any_non_null` and `value_or_null`. Create stubs for corresponding creation reasons.

Default jkinds created from sort variables to `Non_null`. In a future PR, most of those will be updated to accept `Or_null`.

Test those new jkinds.

### CODE REVIEW

File: ocaml/typing/jkind.ml
Comment (goldfirere) [inline ocaml/typing/jkind.ml:1042]: This exported function probably needs to expose the nullability choice to callers.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:406]: Since `immediate` is `mode_crossing`, it's automatically `Non_null`.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:456]: `word` must be `Non_null`. We can't have `nativeint#` have a jkind that's `Or_null`, because then `nativeint# array` would be impossible. Same for `bits32`, `bit64`, `float32`, `float64`.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:400]: In context of this PR, `Non_null` means "can't be a null pointer", and `Or_null` means "can be a null pointer". Under our design constraints, array type arguments must be non-null. Your interpretation would break that constraint. This does not apply specifically to `void`, but I decided to use `Non_null` here for consistency.

We don't have a design for `'a or_null` on non-value layouts, and the `or_null` type I'm adding in subsequent PRs is restricted to values. I think we can decide this question later.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml:99]: That proposal turned out to be independent of the nullability axis. It also needs, at minimum, a lot of changes to module subtyping to be backwards-compatible due to examples like:

```ocaml
module M : sig
  type 'a t
 end = struct
  type 'a t = 'a
 end
 ```

I'm not convinced this proposal can be made backwards-compatible at all, and I'm hesitant to commit to it in a CR. So here I just document the current behavior.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:1042]: This is implemented in the next PR.

Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:1042]: See `(* CR layouts v3.0: split those functions in two versions, one [or_null] and one [non_null]. *)`.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml:1]: Implemented tests for constraints and `type a.`.

File: ocaml/testsuite/tests/typing-layouts/basics_alpha.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts/basics_alpha.ml]: I ended up removing this particular bunch of tests because representability is pretty thoroughly tested in this and next PRs anyway.

File: ocaml/typing/jkind.ml
Comment (goldfirere) [inline ocaml/typing/jkind.ml:369]: I personally find it confusing to include this in `mode_crossing`. It's not really a mode in the same sense as the others. You've convinced me that it can be seen as a bound, but I would not expect to have the `mode_crossing` function affect nullability. Can you move this out from this function?

File: ocaml/typing/jkind.ml
Comment (goldfirere) [reply ocaml/typing/jkind.ml:456]: Something's wrong in the design then. I think the problem in the design is that we're trying to work with `array` at all.

New proposal: restrict `array` to work only with values, but allow `Uniform_array.t` to work with arbitrary layouts. (Not sure necessarily how to implement this given that `Uniform_array.t` is built on top of `array`, but design first, implement second.) I will raise this point internally.

**Comment by dkalinichenko-js** [comment]
Thought: here, I split off `concrete_non_null_creation_reason` from `concrete_creation_reason`. However, this naming doesn't really explain why we need that distinction.

I think we should rename those to something like `concrete_default_creation_reason` and `concrete_permissive_creation_reason` respectively. Before this patch, we always wanted the topmost jkind on the lattice everywhere. However, now we need to default legacy code to `Non_null`, but still accept `Or_null` in as much places as we can.

This will also help us implement the variable syntax: we would just need to change the default in `concrete_default_creation_reason` per file.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml:456]: The separability axis should solve this long-term, as discussed.

**Comment by goldfirere** [comment]
> Thought: here, I split off `concrete_non_null_creation_reason` from `concrete_creation_reason`. However, this naming doesn't really explain why we need that distinction.
> 
> I think we should rename those to something like `concrete_default_creation_reason` and `concrete_permissive_creation_reason` respectively. Before this patch, we always wanted the topmost jkind on the lattice everywhere. However, now we need to default legacy code to `Non_null`, but still accept `Or_null` in as much places as we can.
> 
> This will also help us implement the variable syntax: we would just need to change the default in `concrete_default_creation_reason` per file.

Something seems a little off here, in that I think that the choice between `or_null` and `non_null` happens independently from the choice to be concrete or not. I don't much care about the names of those types, but I don't think this change will accomplish the goal you're describing, unless I'm missing something.

File: ocaml/testsuite/tests/typing-layouts-or-null/basics.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-or-null/basics.ml:340]: OK. Let's keep all these types at `non_null` for now. No reason to hold up this patch.

File: ocaml/typing/jkind.ml
Comment (goldfirere) [reply ocaml/typing/jkind.ml:369]: No -- that is a mode. It's deep!

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-layouts-or-null/variables.ml]: I'm worried about this, too. We probably need a

```
type 'a 'b 'c. 'c constrained' = bool constraint 'c = ('a, 'b) result
```

syntax or something, where we can put definitive kind annotations on the binders at the beginning.

But let's wait to write this syntax until we need it. Constraints are not common.

**Review by goldfirere — APPROVED** [review]
Approved pending resolution of existing discussions.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml]: Why add new syntax instead of changing the behavior of ascriptions to the intuitive one? Needs less changes to support in public release too.

File: ocaml/typing/jkind.ml
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.ml]: Fixed.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml]: Fixed.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml]: Added a test demonstrating that.

File: ocaml/testsuite/tests/typing-layouts-or-null/variables.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-layouts-or-null/variables.ml]: Mostly because I don't know how to do this. Type variables can be born with any kind. But once they're born, they can generally only be constrained to be less than some other kind. I can imagine some pre-pass that looks for ascriptions and has the type variable born at the right kind, but then what do we do with `val f : ('a : any) -> ('a : value)`? If `'a : value`, then `'a : any` is just fine.

**Pull request merged**
